### DISCUSSION TITLE
I need a KeyboardLayout widget example

### DISCUSSION THREAD

**Discussion by s-aleck** [qtile/qtile#3197] [Q&A]
Can't get `KeyboardLayout` to work. I wasn't able to find examples to get this going.

I need a widget which will enable me to switch between US English, Serbian Latin and Serbian Cyrillic (in that order) with `Alt + F4` and display the current layout.

I'd appreciate it if somebody would be so kind as to post their full `KeyboardLayout` configuration.

I'm using Qtile on Arco Linux.

**Comment by m-col** [comment]
I don't use this widget but I might be able to help you figure this out.

Assuming you are using X11, are you able to set the keyboard layouts using `setxkbmap`?

**Reply by s-aleck** [reply]
Thank you. I've  just been experimenting with it.

I copied this from Derek Taylor's config and the widget appears:

>               widget.KeyboardLayout(
>                       foreground = colors[1],
>                       background = colors[8],
>                      fmt = 'Keyboard: {}',
>                      padding = 5
>                       ),

When I try e. g. `setxkbmap es`, the layout is switched to Spanish and the widget reflects that. When I click the widget, the layout is back to US.

However, I can't get the `Alt + Shift` keybinding to work.  That's my problem right now. That and perhaps setting a mouse callback so that the widget toggles the layouts when I click on it.



I commented out all the `Alt + Shift` based shortcuts from `sxhkdrc` and changed `00-keyboard.conf`.
My `localectl` output:

> System Locale: LANG=en_US.UTF-8
>                  LC_NUMERIC=sr_RS
>                  LC_TIME=sr_RS
>                  LC_MONETARY=sr_RS
>                  LC_PAPER=sr_RS
>                  LC_NAME=sr_RS
>                  LC_ADDRESS=sr_RS
>                  LC_TELEPHONE=sr_RS
>                  LC_MEASUREMENT=sr_RS
>                  LC_IDENTIFICATION=sr_RS
>       VC Keymap: us
>      X11 Layout: us,rs
>       X11 Model: pc104
>     X11 Variant: latin
>     X11 Options: grp:alt_shift_toggle

**Reply by m-col** [reply]
If you wanted to toggle between `us` and `es` you would pass in `configured_keyboards=["us", "es"]`, for example.

> However, I can't get the Alt + Shift keybinding to work.

Support for modifier-only keybindings is a bit... serendipitous. If you have any other bindings that use alt and shift, this binding will be executed every time. With that caveat in mind, this might work:
```py
Key(["mod1"], "shift", lazy.widget["keyboardlayout"].next_keyboard(), desc="Next keyboard layout."),
```
See relevant issue: https://github.com/qtile/qtile/issues/2778

**Comment by s-aleck** [comment]
I put this into my `config.py`:

`configured_keyboards = ["us", "rs"],`

(Btw, I wish I could specify the Latin variant of RS here and put all the layouts into a specific order.)

`Key(["mod1", "shift"], "k", lazy.widget["keyboardlayout"].next_keyboard(), desc="Next keyboard layout."),`

It works only once, and I think I know why. When I change the keyboard layout, it affects the keybindings. I'm typing Cyrillic and nothing happens when I try `Super + Shift + Q`, for example. But when I click on the widget and return the layout to US, all the keybindings work properly.

**Reply by m-col** [reply]
Hmm that sounds annoying. I'm afraid that's as far as my help can go, as I only use the one keyboard layout, sorry!

**Comment by s-aleck** [comment]
Found a workaround. Added three keybindings for the following to my `sxhkdrc`:

`setxkbmap -layout us`

`setxkbmap -layout rs -variant latin`

`setxkbmap -layout rs`

Needs a reboot. It works. It's a bit slow, or the widget doesn't display the change right away. Maybe because it's a virtual machine? That slowness is really annoying me.

Aside from the slowness, the keybindings containing letters may stop working when a non-US layout is activated. I think this can be solved by going through all the keybindings and adding alternative keybindings with the special letters which do the exact same thing. Very pedestrian, right? (Imagine if somebody needed more than three layouts.) But right now I see no other way, and I don't think I could use Qtile if I couldn't easily switch between layouts.

EDIT Tried adding such a keybinding, ie. with a special character, to my `sxhkdrc`, but it doesn't work.

**Comment by Pakrohk** [comment]
This is a really big problem and needs to be fixed.
And the strange thing is that if you use something like this!
`setxkbmap -model pc104 -layout ir,us -option grp:win_space_toggle`
Keyboard settings in qtile will work without problems, this  problem only happens if you use the `Keyboard Layout` widget.

**Comment by elParaguayo** [comment]
Thanks for the comment.

The fact that there's a pretty simple workaround means this isn't a big problem as it doesn't render qtile unusable.

We're a small team of developers so we need to focus on things that do break qtile or things that add important functionality.

For less urgent bugs, we'll fix them when we can or, ideally, one of our users will submit a PR.

I hope that makes sense.

**Reply by Pakrohk** [reply]
Thank you. 
you have really developed one of the most amazing window managers, yes there is a solution for it, but it's still a problem because its display is important in the panel and it has a very negative effect on the overall experience of using the window manager.
Think you're using your system, but suddenly your keyboard doesn't work at all. (really annoying)
By the way, I had a strange problem by entering this Bind into my config (x11 was broken).
 I need to comment in the tty to fix it, where does the problem come from?
`
Key([alt], "shift", lazy.widget["keyboardlayout"].next_keyboard()),`

Now I have to change the language from within the panel every time.

**Reply by elParaguayo** [reply]
Is `alt` defined in your config? It's a variable name here.

Also, is `shift` a valid key name? What does `xev` show when you press it?

**Reply by Pakrohk** [reply]
> Is `alt` defined in your config? It's a variable name here.
> 
> Also, is `shift` a valid key name? What does `xev` show when you press it?

ya , alt = "mod1"
Yes, I used shift for my other keys.

**Reply by Pakrohk** [reply]
I solved the problem, but the solution is very strange, I can use `Shift` for the rest of the keys, but `Shift_L` must be used for the layout keyboard.
```py
Key([alt], "Shift_L", lazy.widget["keyboardlayout"].next_keyboard()),
```

**Reply by elParaguayo** [reply]
Are you sure about that? I don't think "shift" is a valid key name. I was expecting it to be `Shift_L/R` as per your code above.

This is in our keysym code:
https://github.com/qtile/qtile/blob/3fb4910f9b780c6eef568c877be37ef14af7114d/libqtile/backend/x11/xkeysyms.py#L360-L361

EDIT: Ah... maybe the confusion is because "shift" is ok for a modifier but not the key name.

**Reply by Pakrohk** [reply]
> Are you sure about that? I don't think "shift" is a valid key name. I was expecting it to be `Shift_L/R` as per your code above.
> 
> This is in our keysym code:
> 
> https://github.com/qtile/qtile/blob/3fb4910f9b780c6eef568c877be37ef14af7114d/libqtile/backend/x11/xkeysyms.py#L360-L361

https://github.com/Pakrohk/qtile/blob/main/cfg/bindings.py

As you can see it was working correctly for me (until now).

**Reply by elParaguayo** [reply]
I think you replied before my edit. Those are all using shift as a modifier rather than the name of key. 

Another example would be using the windows/super key. This is `mod4` when using it as a modifier but, if you wanted to bind something to just the super key it would be
``` python
Key([], "Super_L", ...)
```

**Reply by Pakrohk** [reply]
Well, the question that comes to me now is why when I wrote the keyboard layout , it didn't work like this and Qtile had a problem instead?

**Comment by Pakrohk** [comment]
Another problem I encountered today was that the same problem exists for print screen. I have three keys
`Print, 3270_PrintScreen, 0xFD1D`
I used it and unfortunately `qtile` stopped and sent me to `lightdm`.
What's more strange is that it always used its default configuration when there was a problem in the `qtile` configuration, but now it sends me to `lightdm`.
I used arch repositories to install it.

**Reply by elParaguayo** [reply]
Yes, an error in your config file should result in qtile loading the default config. If qtile crashes completely then you would be returned to your display manager.

There's not really enough here to help debug your issue. Ideally we'd need a link to the broken config. You should also run `python ~/.config/qtile/config.py` (to see if there are any errors in your config) and check the log file for error messages related to the crash.

**Reply by Pakrohk** [reply]
Well, I already sent you the link to my settings.
But about the log that you told me to send, by executing the command
`python ~/.config/qtile/config.py`
  No output is printed in the terminal for me.

**Reply by elParaguayo** [reply]
That's not the log. That's just checking if there are any errors in your config. The logfile is at `~/.local/share/qtile/qtile.log`.

**Reply by Pakrohk** [reply]
> That's not the log. That's just checking if there are any errors in your config. The logfile is at `~/.local/share/qtile/qtile.log`.

Yes, I think you misunderstood what I meant, I meant that no output was displayed in the terminal.
This means that there is no problem in the config.py code, also mypy has similar behavior.

Of course, this is a strange problem, because qtile crashes completely.
Of course, I solved the problem of the keys.

The method that was problematic here:
```python
 Key(["Print"], lazy.spawn(AppName)),
```
The correct way to use the keys:

```py
Key([], "Print", lazy.spawn(AppName)),
```

**Reply by elParaguayo** [reply]
Ah, no. I think you misunderstood me!

Running `python /path/to/config` will only confirm that your config file is valid python (no SyntaxErrors etc). However, it won't check if your keybindings are valid. That only happens when qtile loads and parses your config. Any errors then would be printed to the logfile that I referred to.

`Key(["Print"], lazy.spawn(AppName)),` is valid Python syntax so won't raise any error until the server tries to bind the key.
